storm-client/src/jvm/org/apache/storm/topology/PersistentWindowedBoltExecutor.java

```diff
@@ -0,0 +1,563 @@
+                        throw new IllegalStateException("No calls to next() since last call to remove()");
+                    }
+                    removeFrom.remove();
+                    removeFrom = null;
+                }
+
+                @Override
+                public boolean hasNext() {
+                    boolean curHasNext = current.hasNext();
+                    while (!curHasNext && ids.hasNext()) {
+                        if (curPartition != null) {
+                            unpin(curPartition.getId());
+                        }
+                        curPartition = getPinnedPartition(ids.next());
+                        iteratorPins.add(curPartition.getId());
+                        if (curPartition != null) {
+                            current = curPartition.iterator();
+                            curHasNext = current.hasNext();
+                        }
+                    }
+                    // un-pin the last partition
+                    if (!curHasNext && curPartition != null) {
+                        unpin(curPartition.getId());
+                        curPartition = null;
+                    }
+                    return curHasNext;
+                }
+
+                @Override
+                public Event<T> next() {
+                    if (!hasNext()) {
+                        throw new NoSuchElementException();
+                    }
+                    removeFrom = current;
+                    return current.next();
+                }
+
+                private void unpin(long id) {
+                    cache.unpin(id);
+                    iteratorPins.remove(id);
+                }
+            };
+        }
+
+        void clearIteratorPins() {
+            LOG.debug("clearIteratorPins '{}'", iteratorPins);
+            Iterator<Long> it = iteratorPins.iterator();
+            while (it.hasNext()) {
+                cache.unpin(it.next());
+                it.remove();
+            }
+        }
+
+        @Override
+        public int size() {
+            throw new UnsupportedOperationException();
+        }
+
+        void prepareCommit(long txid) {
+            flush();
+            partitionIds.prepareCommit(txid);
+            windowPartitions.prepareCommit(txid);
+            windowSystemState.prepareCommit(txid);
+        }
+
+        void commit(long txid) {
+            partitionIds.commit(txid);
+            windowPartitions.commit(txid);
+            windowSystemState.commit(txid);
+        }
+
+        void rollback() {
+            partitionIds.rollback();
+            windowPartitions.rollback();
+            windowSystemState.rollback();
+        }
+
+        private void initPartitions() {
+            pids = partitionIds.get(PARTITION_IDS_KEY, new LinkedList<>());
+            if (pids.isEmpty()) {
+                pids.add(0L);
+                partitionIds.put(PARTITION_IDS_KEY, pids);
+            } else {
+                latestPartitionId = pids.peekLast();
```

Comment:
yes, it was initialized to 0 being a member variable, anyways I will remove the else to make it clearer.